### DISCUSSION TITLE
JPypeDemoDoesn'tRun

### DISCUSSION THREAD

**Discussion by vsquared** [py5coding/py5generator#471] [Q&A] [answered]
Anybody know how to fix this; it's their code:
```python
import jpype
import jpype.imports

jpype.startJVM()
import java
import javax
from javax.swing import *

def createAndShowGUI():
    print("Got to here.")
    frame = JFrame("HelloWorldSwing")
    print("Never gets to here.")
    frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE)
    frame.setBounds(100, 100, 400, 400)
    label = JLabel("Hello World")
    frame.getContentPane().add(label)
    frame.pack()
    frame.setVisible(True)

# Start an event loop thread to handling gui events
@jpype.JImplements(java.lang.Runnable)
class Launch:
    @jpype.JOverride
    def run(self):
        createAndShowGUI()
javax.swing.SwingUtilities.invokeLater(Launch())

```

**Comment by villares** [comment]
Cheers @vsquared ! I'm the one who insisted you should try to learn how to make the javax GUI stuff work with "pure jpype", and I really think finding out a way of making this work is important, but I'm afraid this is not the best discussion forum for you to find help with this issue! Here, *you* are the most knowledgeable person on `javax` & `JFrame` & etc. no one else on the py5 forum knows much about it. I think you should post this on some jpype discussion list or forum where you'll have a much better chance of finding people who can help you, don't you agree?

**Reply by vsquared** [reply]
Fair enough.  I thought since you utilized jpype for py5 you might know how to fix it.  I think it's odd that they post something which doesn't seem to run, so they should probably be the ones to fix it (if indeed it's their issue and not mine or my system).

**Addendum:**
Looks like somebody else had the same problem four years ago: https://github.com/jpype-project/jpype/issues/906  I'm not sure it ever got solved.

**Reply by vsquared** [reply]
It appears your collaborator knows all about this: https://github.com/py5coding/py5generator/issues/4  Are you using pyobjc behind the curtain?

**Reply by villares** [reply]
Yes, it is true Jim will known about that, but I'm afraid this could overwhelm him with something a bit off-topic. Maybe I'm being over-reactive and over-protective..., and I wouldn't want to answer for him. I hope you understand I really enjoy your explorations, I want to help, but I also want to promote the right set of expectations for the py5 forum :)

**Reply by hx2A** [reply]
@vsquared , the maintainer of the jpype project is very diligent and responsive. He knows Java and JPype very well and I believe he'd be able to help you with this.

> Yes, it is true Jim will known about that, but I'm afraid this could overwhelm him with something a bit off-topic. Maybe I'm being over-reactive and over-protective..., and I wouldn't want to answer for him. I hope you understand I really enjoy your explorations, I want to help, but I also want to promote the right set of expectations for the py5 forum :)

Thank you, @villares !

> Are you using pyobjc behind the curtain?

py5 does use pyobjc on MacOS computers. the pyobjc library is listed as one of the dependencies and is installed along with py5 on MacOS computers. pyobjc is used when you use py5 outside of Jupyter with the generic python interpreter or with the run_sketch utility (what Thonny uses).

I do recall looking at the Jpype code and there is something in there for an event loop on MacOS computers. I vaguely remember reading that it was a PR that somebody made and that it probably needed work. It isn't mentioned in the JPype documentation. That didn't work for py5 but I did study it carefully and the code did inspire some of the ideas used for what py5 does with pyobjc.

**Reply by vsquared** [reply]
I did get a response which basically says that I need to get the AWT EventDispatchThread to work in Python.  As far as I can tell no one has ever done that for JPype; I did see some code for Jython (which I have no experience with).  I think it would be nice if he placed a disclaimer in the Documentation stating that the demo doesn't work on a mac.  I wouldn't mind looking at the JPype source code to look at that event loop, but I'm tiring of hitting dead ends with this so may soon decide to move on.

I just thought it was interesting that you used pyobjc under the hood.  I did a few demos in pyobjc before your friend convinced me to come over here, and I was impressed how well it ran in Thonny once I got the syntax right.  I found one other guy willing to help me a little, but other than that it appeared to be a seldom used language, so I never posted the demos.  I have a rather extensive background in objc using the programmatic approach so it was fairly easy to cross over.  I know how to use Xcode but for small stuff I use homemade editors for objc and a little swift.  More complex work goes into Xcode.

I'm getting ready to turn loose of a rudimentary py5 editor that in no way will replace Thonny, but I'd be interested to hear what you think of it.

**Reply by hx2A** [reply]
> I just thought it was interesting that you used pyobjc under the hood.

It took many months for me to figure out how to get py5 to work correctly on MacOS. Initially there were many serious bugs. Eventually I figured out how to get it to work in Jupyter notebooks and how to get it to work in the generic Python interpreter.

> I'm getting ready to turn loose of a rudimentary py5 editor that in no way will replace Thonny, but I'd be interested to hear what you think of it.

Sure, sounds interesting!

**Comment by vsquared** [comment, marked as answer]
AppHelper is the key to running this on a Mac:
```python
import jpype
import jpype.imports

jpype.startJVM()

import java
import javax
from javax.swing import *
from PyObjCTools import AppHelper

def buildWnd():
    print("buildWnd called.")
    print("frame in buildWnd =",frame)
    btn = javax.swing.JButton("Button")
    btn.setBounds(30,60,100,24)
    frame.add(btn)
    frame.setVisible(True)
    print("btn =",btn)
    print("EDT =",javax.swing.SwingUtilities.isEventDispatchThread())
    print("Done in buildWnd.")
    
def main():
    global frame

    frame = javax.swing.JFrame("Swing EDT")
    frame.setDefaultCloseOperation(javax.swing.JFrame.EXIT_ON_CLOSE)
    frame.setBounds(100,100,400,400)
    frame.setLayout(None)

    javax.swing.SwingUtilities.invokeLater(buildWnd)
    AppHelper.runEventLoop()

main()
```
**Output:**
<img width="405" alt="helper" src="https://github.com/py5coding/py5generator/assets/2294268/8ac83c11-ba61-4350-b24d-cbb5fd5a77cf">

**Addendum:**
This also works:
```python
import jpype
import jpype.imports

jpype.startJVM()
import java
import javax
from javax.swing import *
from Cocoa import NSApp

def createAndShowGUI():
    frame = JFrame("HelloWorldSwing")
    frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE)
    label = JLabel("Hello World")
    frame.getContentPane().add(label)
    frame.pack()
    frame.setVisible(True)

javax.swing.SwingUtilities.invokeLater(createAndShowGUI)
NSApp.run()

```
The key to running this on a Mac is the last line of code which comes from the Cocoa framework, made possible by the PyObjc plugin to Thonny.

**Addendum2:**
This should make it cross-platform.  @villares @hx2A  Please see if you can run this on your systems.  Thanks.
```python
import jpype
import jpype.imports

jpype.startJVM()
import java
import javax
import platform
from javax.swing import *
if(platform.system() == "Darwin"):
    from Cocoa import NSApp

def createAndShowGUI():
    frame = JFrame("HelloWorldSwing")
    frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE)
    label = JLabel("Hello World")
    frame.getContentPane().add(label)
    frame.pack()
    frame.setVisible(True)

javax.swing.SwingUtilities.invokeLater(createAndShowGUI)
print(platform.system())
if(platform.system() == "Darwin"):    
    NSApp.run()
```

**Comment by vsquared** [comment]
@villares This is the only other demo that needs testing.  I think it works cross-platform, but the author of JPype is not happy with it because he wants to start the main event loop up at the top and that is just not the 'Apple way' in my experience.

**Reply by villares** [reply]
This kind of worked on Windows, the problem was that the window got stuck in the top left corner of the screen. It could be maximized and restored, and the had to use the corner java logo/icon menu on the window and the keyboard to move it (couldn't drag it with the mouse).  
![image](https://github.com/py5coding/py5generator/assets/3694604/a75d28a6-225b-47c9-976f-215e5d235452)

**Reply by villares** [reply]
I'll try on Linux when I get home.

**Reply by vsquared** [reply]
It's not a very good demo and is just a little window stuck in the left upper corner of the screen.  The window should be resizable.  The important thing is that it runs and you can see the window.  Thanks for your help.

**Reply by villares** [reply]
Works on Linux too!
![image](https://github.com/py5coding/py5generator/assets/3694604/e1002ddb-e530-4ff7-8159-83f46c4761c7)

**Reply by vsquared** [reply]
Thanks for telling me about this approach.  The code that is on the JPype website for AWT/Swing won't run on a Mac (ie, no output to screen), but after making a couple of changes it will produce screen output, and you've just proven that the same code still runs on Windows and Linux.  That makes it cross-platform in my book.  Thanks for testing.
